File: Sources/Shared/Coordinators/AppNavigationCoordinator.swift
// MARK: - AppNavigationCoordinator
// Type-safe @Observable replacement for cross-component NotificationCenter events.
// macOS 14+, Swift 5.10

import Foundation
import Observation

enum AppMode: Equatable {
    case regular
    case codeSpeak
}

/// Coordinates global navigation actions (install wizard, settings) between
/// the toolbar (sender) and the root view (receiver).
///
/// Replaces two untyped `NotificationCenter` posts:
///  - `.showInstallAgentWizard` → `agentToInstall: AIAssistant?`
///  - `.showAppSettings`        → `showingSettings: Bool`
///
/// Both `ToolbarView` (in `WindowToolbarRemover`) and `RootView` receive this
/// coordinator via `@Environment(\.navigationCoordinator)` — no hidden coupling.
@Observable
@MainActor
final class AppNavigationCoordinator {

    /// The agent whose install wizard should be shown. Set to `nil` to dismiss.
    var agentToInstall: AIAssistant?

    var currentMode: AppMode = .regular

    var codeSpeakBuildRequested: Bool = false

    // MARK: - CodeSpeak Run Bar State
    // Extracted to `CodeSpeakRunBarState` to respect SRP.

    /// Run bar state: command selection, inputs, running/stop flags, editor state.
    var runBar = CodeSpeakRunBarState()

    // MARK: - CodeSpeak Titlebar Layout

    /// Width of the CodeSpeak specs sidebar column (left panel).
    ///
    /// Updated by `CodeSpeakModeView` via `GeometryReader` so `ToolbarView` can
    /// position the breadcrumb exactly above the center column's left edge.
    var specsColumnWidth: CGFloat = 220

    /// Update `currentMode` based on whether the active project is a CodeSpeak project.
    ///
    /// Called by `AppLifecycleCoordinator` whenever the active project changes.
    /// The mode is determined solely by the presence of `codespeak.json` at
    /// the project root — no user toggle.
    func syncMode(isCodeSpeak: Bool) {
        let newMode: AppMode = isCodeSpeak ? .codeSpeak : .regular
        guard newMode != currentMode else { return }
        currentMode = newMode
        if newMode == .codeSpeak {
            showingChangesPanel = false
            showingSpecPanel = false
            showingTraceabilityPanel = false
        }
    }

    /// Set to `true` to open the Settings window.
    ///
    /// `RootView` observes this and calls `openSettings()`, then resets to `false`.
    var showingSettings: Bool = false

    /// Controls the right-side git changes panel visibility.
    ///
    /// Toggled from the toolbar button (sidebar.right icon) or ⌘⇧G shortcut.
    /// `RootView` observes this to conditionally render `GitChangesPanelView`.
    var showingChangesPanel: Bool = false

    /// Controls the right-side CodeSpeak spec build panel visibility.
    ///
    /// Toggled via ⌘⇧S shortcut.
    /// `RootView` observes this to conditionally render `SpecBuildPanelView`.
    var showingSpecPanel: Bool = false

    /// Controls the right-side traceability panel visibility.
    ///
    /// Toggled via toolbar button. `RootView` observes this.
    var showingTraceabilityPanel: Bool = false

}

File: shared/src/commonMain/kotlin/studio/vibe/shared/coordinator/AppNavigationCoordinator.kt
package studio.vibe.shared.coordinator

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import studio.vibe.shared.contract.AIAgent
import studio.vibe.shared.model.CodeSpeakRunBarState

/**
 * Application mode — determines which UI layout is active.
 */
enum class AppMode {
    Regular,
    CodeSpeak,
}

/**
 * Coordinates global navigation actions (install wizard, settings, panel visibility)
 * between the toolbar and root view layers.
 *
 * Replaces untyped NotificationCenter posts with typed StateFlow properties that
 * any component can observe independently.
 */
class AppNavigationCoordinator {

    // MARK: - Agent Install Wizard

    /** The agent whose install wizard should be shown. `null` = dismissed. */
    private val _agentToInstall = MutableStateFlow<AIAgent?>(null)
    val agentToInstall: StateFlow<AIAgent?> = _agentToInstall.asStateFlow()

    fun setAgentToInstall(agent: AIAgent?) {
        _agentToInstall.value = agent
    }

    // MARK: - App Mode

    private val _currentMode = MutableStateFlow(AppMode.Regular)
    val currentMode: StateFlow<AppMode> = _currentMode.asStateFlow()

    // MARK: - CodeSpeak Build

    private val _codeSpeakBuildRequested = MutableStateFlow(false)
    val codeSpeakBuildRequested: StateFlow<Boolean> = _codeSpeakBuildRequested.asStateFlow()

    fun setCodeSpeakBuildRequested(requested: Boolean) {
        _codeSpeakBuildRequested.value = requested
    }

    // MARK: - CodeSpeak Run Bar State

    private val _runBar = MutableStateFlow(CodeSpeakRunBarState())
    val runBar: StateFlow<CodeSpeakRunBarState> = _runBar.asStateFlow()

    fun updateRunBar(mutate: (CodeSpeakRunBarState) -> CodeSpeakRunBarState) {
        _runBar.update(mutate)
    }

    // MARK: - Settings

    private val _showingSettings = MutableStateFlow(false)
    val showingSettings: StateFlow<Boolean> = _showingSettings.asStateFlow()

    fun setShowingSettings(showing: Boolean) {
        _showingSettings.value = showing
    }

    // MARK: - Panels

    private val _showingChangesPanel = MutableStateFlow(false)
    val showingChangesPanel: StateFlow<Boolean> = _showingChangesPanel.asStateFlow()

    fun setShowingChangesPanel(showing: Boolean) {
        _showingChangesPanel.value = showing
    }

    private val _showingSpecPanel = MutableStateFlow(false)
    val showingSpecPanel: StateFlow<Boolean> = _showingSpecPanel.asStateFlow()

    fun setShowingSpecPanel(showing: Boolean) {
        _showingSpecPanel.value = showing
    }

    private val _showingTraceabilityPanel = MutableStateFlow(false)
    val showingTraceabilityPanel: StateFlow<Boolean> = _showingTraceabilityPanel.asStateFlow()

    fun setShowingTraceabilityPanel(showing: Boolean) {
        _showingTraceabilityPanel.value = showing
    }

    // MARK: - Sidebar

    private val _showingSidebar = MutableStateFlow(true)
    val showingSidebar: StateFlow<Boolean> = _showingSidebar.asStateFlow()

    fun setShowingSidebar(showing: Boolean) {
        _showingSidebar.value = showing
    }

    fun toggleSidebar() {
        _showingSidebar.update { !it }
    }

    // MARK: - CodeSpeak Column Widths (toolbar three-section layout)

    /**
     * Width of the specs-list column in CodeSpeak mode, in pixels (not dp).
     * Updated by [SpecsListColumn] via GeometryReader equivalent (onGloballyPositioned).
     * Read by ToolbarView Box1 to align the breadcrumb to the centre column.
     */
    private val _specsColumnWidth = MutableStateFlow(220f)
    val specsColumnWidth: StateFlow<Float> = _specsColumnWidth.asStateFlow()

    fun setSpecsColumnWidth(widthPx: Float) {
        if (widthPx > 0f) _specsColumnWidth.value = widthPx
    }

    // MARK: - Mode Sync

    /**
     * Update [currentMode] based on whether the active project is a CodeSpeak project.
     *
     * Called by the app lifecycle coordinator whenever the active project changes.
     * The mode is determined solely by the presence of `codespeak.json` at the
     * project root — no user toggle. Switching away from CodeSpeak resets all panels.
     */
    fun syncMode(isCodeSpeak: Boolean) {
        val newMode = if (isCodeSpeak) AppMode.CodeSpeak else AppMode.Regular
        if (newMode == _currentMode.value) return
        _currentMode.value = newMode
        if (newMode == AppMode.CodeSpeak) {
            _showingChangesPanel.value = false
            _showingSpecPanel.value = false
            _showingTraceabilityPanel.value = false
        }
    }
}
